src/Dialect/ONNX/ONNXOps/Tensor/Squeeze.cpp

```diff
@@ -183,3 +183,39 @@ namespace onnx_mlir {
 template struct ONNXCommonSqueezeOpShapeHelper<ONNXSqueezeOp>;
 template struct ONNXCommonSqueezeOpShapeHelper<ONNXSqueezeV11Op>;
 } // namespace onnx_mlir
+
+//===----------------------------------------------------------------------===//
+// Folder
+//===----------------------------------------------------------------------===//
+OpFoldResult ONNXSqueezeOp::fold(FoldAdaptor adaptor) {
+  OnnxElementsAttrBuilder elementsBuilder(getContext());
+  if (!adaptor.getData() || !adaptor.getAxes()) {
+    // Use original Op if Data or Axes is not constant
+    return nullptr;
+  }
+
+  // ToFix: copied from constprop.cpp.
```

Comment:
this is the way to do it and there's nothing to fix, right? i.e. we shouldn't have "ToFix" in the comment

if the result shape is not static then we just cannot fold, because the constant folding breaks down if a constant op has non-static shape, as it must derive its type from the attribute, which always has static type

Reply:
I am thinking of calling Shape inference here to fix it.

Reply:
the issue is that if we change the operation's result type then things go hay-wire, especially if it's the last op in a function body then MLIR will complain that the return result (operand of the return op) doesn't match the function's return type, so we will need to propagate the type change to some of the users of the return value, which might have knock-on effects wherever we canonicalize ops

maybe we should do that: change the canonicalize pass etc to do the necessary shape inference and shape propagation to support constant folding, but I can't predict how many obstacles we'll encounter if we try that

Reply:
I tried to add the shape inference. It works for my test case. I will check the last op issue you mentioned.
Now `onnx-mlir-opt --canonicalize` of the following model (in onnx_canonicalization.mlir)
```
func.func @test_fuse_add_conv(%arg0 : tensor<1x1x28x28xf32>, %arg1 : tensor<8x1x5x5xf32>) -> tensor<1x8x28x28xf32> {
    %cst = "onnx.NoValue"() {value} : () -> none
    %0 = "onnx.Conv"(%arg0, %arg1, %cst) {auto_pad = "SAME_UPPER", dilations = [1, 1], group = 1 : si64, kernel_shape = [5, 5], onnx_node_name = "Convolution28", strides = [1, 1]} : (tensor<1x1x28x28xf32>, tensor<8x1x5x5xf32>, none) -> tensor<1x8x28x28xf32>
    %1 = onnx.Constant dense<[[[-0.161539719]], [[-0.433835655]], [[0.091641359]], [[-0.0168522168]], [[-0.0650264397]], [[-0.131737873]], [[0.0204175506]], [[-0.121110231]]]> : tensor<8x1x1xf32>
    %2 = "onnx.Add"(%0, %1) : (tensor<1x8x28x28xf32>, tensor<8x1x1xf32>) -> tensor<1x8x28x28xf32> 
    return %2 : tensor<1x8x28x28xf32> 
}
```
Will generate the result 
```
module {
  func.func @test_fuse_add_conv(%arg0: tensor<1x1x28x28xf32>, %arg1: tensor<8x1x5x5xf32>) -> tensor<1x8x28x28xf32> {
    %0 = onnx.Constant dense<[-0.161539719, -0.433835655, 0.091641359, -0.0168522168, -0.0650264397, -0.131737873, 0.0204175506, -0.121110231]> : tensor<8xf32>
    %1 = "onnx.Conv"(%arg0, %arg1, %0) {auto_pad = "SAME_UPPER", dilations = [1, 1], group = 1 : si64, kernel_shape = [5, 5], strides = [1, 1]} : (tensor<1x1x28x28xf32>, tensor<8x1x5x5xf32>, tensor<8xf32>) -> tensor<1x8x28x28xf32>
    return %1 : tensor<1x8x28x28xf32>
  }
}
```
Same as `onnx-mlir-opt --canonicalize --shape-inference --constprop-onnx` with implicit pass ordering removed.